src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -235,180 +276,215 @@ protected override AuthenticationTicket AuthenticateCore()
+                    var securityTokenReceivedNotification =
+                        new SecurityTokenReceivedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+                        {
+                            ProtocolMessage = message
+                        };
+
+                    await Options.Notifications.SecurityTokenReceived(securityTokenReceivedNotification);
+                    if (securityTokenReceivedNotification.HandledResponse)
                     {
-                        validationParameters.ValidIssuer = _configuration.Issuer;
+                        _logger.LogInformation(message: Resources.OIDCH_0008_SecurityTokenReceivedNotificationHandledResponse);
+                        return securityTokenReceivedNotification.AuthenticationTicket;
                     }
-                    else if (!string.IsNullOrWhiteSpace(_configuration.Issuer))
+
+                    if (securityTokenReceivedNotification.Skipped)
                     {
-                        validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? new[] { _configuration.Issuer } : validationParameters.ValidIssuers.Concat(new[] { _configuration.Issuer }));
+                        _logger.LogInformation(message: Resources.OIDCH_0009_SecurityTokenReceivedNotificationSkipped);
+                        return null;
                     }
 
-                    validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? _configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(_configuration.SigningKeys));
-                }
+                    // Copy and augment to avoid cross request race conditions for updated configurations.
+                    TokenValidationParameters validationParameters = Options.TokenValidationParameters.Clone();
+                    if (_configuration != null)
+                    {
+                        if (string.IsNullOrWhiteSpace(validationParameters.ValidIssuer))
+                        {
+                            validationParameters.ValidIssuer = _configuration.Issuer;
+                        }
+                        else if (!string.IsNullOrWhiteSpace(_configuration.Issuer))
+                        {
+                            validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? new[] { _configuration.Issuer } : validationParameters.ValidIssuers.Concat(new[] { _configuration.Issuer }));
+                        }
 
-                AuthenticationTicket ticket;
-                SecurityToken validatedToken = null;
-                ClaimsPrincipal principal = null;
-                JwtSecurityToken jwt = null;
+                        validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? _configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(_configuration.SigningKeys));
```

Comment:
I have not issue with people doing this, but I don't see any reason to change right now.